core/src/main/java/org/kiwix/kiwixmobile/core/history/HistoryActivity.kt

```diff
@@ -0,0 +1,173 @@
+import androidx.appcompat.view.ActionMode
+import androidx.appcompat.view.ActionMode.Callback
+import androidx.appcompat.widget.SearchView
+import androidx.lifecycle.Observer
+import androidx.lifecycle.ViewModelProvider
+import androidx.recyclerview.widget.LinearLayoutManager
+import androidx.recyclerview.widget.RecyclerView
+import io.reactivex.disposables.CompositeDisposable
+import kotlinx.android.synthetic.main.activity_history.history_switch
+import kotlinx.android.synthetic.main.activity_history.no_history
+import kotlinx.android.synthetic.main.activity_history.recycler_view
+import kotlinx.android.synthetic.main.layout_toolbar.toolbar
+import org.kiwix.kiwixmobile.core.R
+import org.kiwix.kiwixmobile.core.R.id
+import org.kiwix.kiwixmobile.core.R.string
+import org.kiwix.kiwixmobile.core.base.BaseActivity
+import org.kiwix.kiwixmobile.core.di.components.CoreComponent
+import org.kiwix.kiwixmobile.core.extensions.ActivityExtensions.coreActivityComponent
+import org.kiwix.kiwixmobile.core.extensions.ActivityExtensions.viewModel
+import org.kiwix.kiwixmobile.core.history.adapter.HistoryAdapter
+import org.kiwix.kiwixmobile.core.history.adapter.HistoryAdapter.OnItemClickListener
+import org.kiwix.kiwixmobile.core.history.adapter.HistoryDelegate.HistoryDateDelegate
+import org.kiwix.kiwixmobile.core.history.adapter.HistoryDelegate.HistoryItemDelegate
+import org.kiwix.kiwixmobile.core.history.adapter.HistoryListItem.HistoryItem
+import org.kiwix.kiwixmobile.core.history.viewmodel.Action
+import org.kiwix.kiwixmobile.core.history.viewmodel.Action.ExitHistory
+import org.kiwix.kiwixmobile.core.history.viewmodel.Action.Filter
+import org.kiwix.kiwixmobile.core.history.viewmodel.Action.OnItemLongClick
+import org.kiwix.kiwixmobile.core.history.viewmodel.Action.RequestDeleteAllHistoryItems
+import org.kiwix.kiwixmobile.core.history.viewmodel.Action.RequestDeleteSelectedHistoryItems
+import org.kiwix.kiwixmobile.core.history.viewmodel.Action.ToggleShowHistoryFromAllBooks
+import org.kiwix.kiwixmobile.core.history.viewmodel.HistoryViewModel
+import org.kiwix.kiwixmobile.core.history.viewmodel.State
+import org.kiwix.kiwixmobile.core.history.viewmodel.State.NoResults
+import org.kiwix.kiwixmobile.core.history.viewmodel.State.Results
+import org.kiwix.kiwixmobile.core.history.viewmodel.State.SelectionResults
+import org.kiwix.kiwixmobile.core.utils.DialogShower
+import org.kiwix.kiwixmobile.core.utils.SimpleTextListener
+import javax.inject.Inject
+
+const val USER_CLEARED_HISTORY: String = "user_cleared_history"
+
+class HistoryActivity : OnItemClickListener, BaseActivity() {
+  private val activityComponent by lazy { coreActivityComponent }
+  private var actionMode: ActionMode? = null
+  @Inject lateinit var viewModelFactory: ViewModelProvider.Factory
+  private val historyViewModel by lazy { viewModel<HistoryViewModel>(viewModelFactory) }
+  private val compositeDisposable = CompositeDisposable()
+
+  @Inject lateinit var dialogShower: DialogShower
```

Comment:
this can be deleted now, right?

Reply:
I keep it in the activity and send it to the historyViewModel when the delete button is clicked. 
```
      override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
        if (item.itemId == id.menu_context_delete) {
          historyViewModel.actions.offer(RequestDeleteSelectedHistoryItems(dialogShower))
          ...
```
I should probably read about dagger as I keep getting in trouble when trying to inject things like the dialogShower elsewhere. When I tried putting the dialogShower in the historyViewModel, the BookDatabaseEntity complained in the build step because of a missing binding.

Reply:
Because a DialogShower requires an Activity in its constructor so you need to call inject on an `activityComponent` for dagger to be able to create it in a class

You can see this behaviour in `ShowDeleteSearchDialog`. It is not the best because this side effect is hardcoded to that Activity `(activity as SearchActivity)` because I never came up with a generic way of retrieving an activitycomponent because there hadn't been a need to do so. And it gets slightly complicated with the modularisation. Regardless you can repeat the pattern.

Important to note we are retrieving the activityComponent, not creating a new one.